File: backend/api/routes/vehicles.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from backend.database.connection import get_db
from backend.database.repositories.vehicle_repository import VehicleRepository
from backend.database.repositories.telemetry_repository import TelemetryRepository
from backend.api.schemas.vehicle import VehicleResponse, VehicleCreate, VehicleDetail, VehicleWithHealth, AnomalyCause
from backend.api.schemas.vehicle import FleetAnomalyType
router = APIRouter(prefix="/vehicles", tags=["Vehicles"])


@router.get("/", response_model=list[VehicleResponse])
def list_vehicles(db: Session = Depends(get_db)):
    repo = VehicleRepository(db)
    return repo.get_all()


@router.get("/with-health", response_model=list[VehicleWithHealth])
def list_vehicles_with_health(db: Session = Depends(get_db)):
    repo = VehicleRepository(db)
    return repo.get_all_with_health()


@router.get("/{vehicle_id}", response_model=VehicleDetail)
def get_vehicle(vehicle_id: int, db: Session = Depends(get_db)):
    vehicle_repo = VehicleRepository(db)
    vehicle = vehicle_repo.get_by_id(vehicle_id)
    if not vehicle:
        raise HTTPException(status_code=404, detail="Vehicle not found")

    telemetry_repo = TelemetryRepository(db)
    total_records = telemetry_repo.count_by_vehicle(vehicle_id)

    return {
        "id": vehicle.id,
        "vehicle_code": vehicle.vehicle_code,
        "model": vehicle.model,
        "created_at": vehicle.created_at,
        "total_records": total_records,
    }


@router.get("/{vehicle_id}/anomaly-causes", response_model=list[AnomalyCause])
def get_anomaly_causes(vehicle_id: int, db: Session = Depends(get_db)):
    vehicle_repo = VehicleRepository(db)
    vehicle = vehicle_repo.get_by_id(vehicle_id)
    if not vehicle:
        raise HTTPException(status_code=404, detail="Vehicle not found")

    telemetry_repo = TelemetryRepository(db)
    return telemetry_repo.get_top_anomaly_causes(vehicle_id)


@router.post("/", response_model=VehicleResponse, status_code=201)
def create_vehicle(payload: VehicleCreate, db: Session = Depends(get_db)):
    repo = VehicleRepository(db)
    return repo.create(vehicle_code=payload.vehicle_code, model=payload.model)


@router.get("/fleet/anomaly-types", response_model=list[FleetAnomalyType])
def get_fleet_anomaly_types(db: Session = Depends(get_db)):
    telemetry_repo = TelemetryRepository(db)
    counts = telemetry_repo.get_fleet_top_anomaly_types()
    return [{"sensor_label": k, "count": v} for k, v in sorted(counts.items(), key=lambda x: -x[1])]
from backend.api.schemas.vehicle import HealthTrendPoint

@router.get("/{vehicle_id}/health-trend", response_model=list[HealthTrendPoint])
def get_health_trend(vehicle_id: int, db: Session = Depends(get_db)):
    vehicle_repo = VehicleRepository(db)
    vehicle = vehicle_repo.get_by_id(vehicle_id)
    if not vehicle:
        raise HTTPException(status_code=404, detail="Vehicle not found")
    telemetry_repo = TelemetryRepository(db)
    return telemetry_repo.get_health_trend(vehicle_id)

File: frontend/src/api/vehicles.ts
import { apiClient } from "./client";
import type { Vehicle, VehicleDetail, PaginatedTelemetry, TelemetryStats, HealthScore } from "../types/vehicle";

export async function getVehicles(): Promise<Vehicle[]> {
  const response = await apiClient.get<Vehicle[]>("/vehicles/");
  return response.data;
}

export async function getVehicleDetail(vehicleId: number): Promise<VehicleDetail> {
  const response = await apiClient.get<VehicleDetail>(`/vehicles/${vehicleId}`);
  return response.data;
}

export async function getVehicleTelemetry(
  vehicleId: number,
  page: number = 1,
  pageSize: number = 100
): Promise<PaginatedTelemetry> {
  const response = await apiClient.get<PaginatedTelemetry>(
    `/vehicles/${vehicleId}/telemetry`,
    { params: { page, page_size: pageSize } }
  );
  return response.data;
}

export async function getVehicleStats(vehicleId: number): Promise<TelemetryStats> {
  const response = await apiClient.get<TelemetryStats>(`/vehicles/${vehicleId}/stats`);
  return response.data;
}

export async function getVehicleHealthScore(vehicleId: number): Promise<HealthScore> {
  const response = await apiClient.get<HealthScore>(`/vehicles/${vehicleId}/health-score`);
  return response.data;
}
export interface VehicleWithHealth {
  id: number;
  vehicle_code: string;
  model: string | null;
  total_records: number;
  anomaly_count: number;
  anomaly_rate: number;
  health_score: number;
}

export async function getVehiclesWithHealth(): Promise<VehicleWithHealth[]> {
  const response = await apiClient.get<VehicleWithHealth[]>("/vehicles/with-health");
  return response.data;
}
export interface AnomalyCause {
  timestamp: string;
  sensor: string;
  sensor_label: string;
  value: number | null;
  deviation: number;
}

export async function getAnomalyCauses(vehicleId: number): Promise<AnomalyCause[]> {
  const response = await apiClient.get<AnomalyCause[]>(`/vehicles/${vehicleId}/anomaly-causes`);
  return response.data;
}
export interface AnomalyCause {
  timestamp: string;
  sensor: string;
  sensor_label: string;
  value: number | null;
  deviation: number;
}

export interface FleetAnomalyType {
  sensor_label: string;
  count: number;
}

export async function getFleetAnomalyTypes(): Promise<FleetAnomalyType[]> {
  const response = await apiClient.get<FleetAnomalyType[]>("/vehicles/fleet/anomaly-types");
  return response.data;
}
export interface HealthTrendPoint {
  date: string;
  total_records: number;
  anomaly_count: number;
  anomaly_rate: number;
  health_score: number;
}

export async function getHealthTrend(vehicleId: number): Promise<HealthTrendPoint[]> {
  const response = await apiClient.get<HealthTrendPoint[]>(`/vehicles/${vehicleId}/health-trend`);
  return response.data;
}